Add User Experience docs (#187)
elastic/observability-docs: docs/en/observability/user-experience.asciidoc

// These need to move into the docs repo
:user-experience: User Experience

> :user-experience-tab: {user-experience} tab

Author: Added in https://github.com/elastic/docs/pull/1980

// At some point we need to add a page that talks about how synthetics and
// RUM/UE work together. Passive/Active, etc.

[[user-experience]]
= {user-experience}

{user-experience} provides a way to quantify and analyze the perceived performance of your web application.
Unlike testing environments, {user-experience} data reflects actual users' real-world experiences.

> With just a few lines of code, you can surface key user experience metrics and investigative tools

Author: In general, I try to avoid linking away in an intro paragraph -- someone just started reading here, and we don't want to send them elsewhere quite yet. With that being said, you're right, this could be more clear. I'll update the text.

In short, a good {user-experience} keeps your users happy and improves your website's odds of success.

[discrete]
[[how-user-experience-works]]
== How does {user-experience} work?

// I'm not even sure if this section is necessary
// Some of this is taken from the RUM docs

{user-experience} metrics are powered by the {apm-rum-ref}[APM Real User Monitoring (RUM) agent].
The RUM agent is installed as a dependency to your application,
and it only takes a few lines of code to <<instrument-apps,get started>>.

The RUM agent uses browser timing APIs, like https://w3c.github.io/navigation-timing/[Navigation Timing],
https://w3c.github.io/resource-timing/[Resource Timing], https://w3c.github.io/paint-timing/[Paint Timing],
and https://w3c.github.io/user-timing/[User Timing], to capture {user-experience}
metrics every time a user hits one of your pages.
This data is stored in {es}, where it can be visualized using {kib}.

[discrete]
[[user-experience-tab]]
== {user-experience} in {kib}

// Need some kind of app introduction here
// Not quite sure yet...
// ...any ideas?

[discrete]
[[user-experience-page-load]]

> === Page load duration

Author: I played with the text a little bit. Do you think these questions should just go away entirely?

> === Page load duration

Reviewer: No, my thought was can we move the questions somewhere higher as having them under page load duration seems like we are trying to define what is page-load but the sentences say something different. Does that make sense?

How long is my server taking to respond to requests?
How much time is spent parsing and painting that content?
How many page views has my site received?
This high-level overview is your analysis starting point.
You won't be able to fix any problems from viewing these metrics alone,
but you'll get a sense of the big picture as you dive deeper into your data.

[role="screenshot"]
image::images/page-load-duration.png[User experience page load duration metrics]

[discrete]
[[user-experience-metrics]]
=== {user-experience} metrics

{user-experience} metrics help you understand the perceived performance of your website.
For example, first contentful paint is the timestamp when the browser begins rendering content.
In other words, it's around this time that a user first gets feedback that the page is actually loading.

[role="screenshot"]
image::images/user-exp-metrics.png[User experience metrics]

// This is collapsed by default
[%collapsible]

> .Metric reference

Author: Great idea!

====
First contentful paint::
The timestamp when the browser begins rendering content (images, text, etc.) from the DOM.
This is when the user can tell that the page is loading.
footnote:[Source and more info: https://developer.mozilla.org/en-US/docs/Glossary/First_contentful_paint[developer.mozilla.org]]

Total blocking time::
The total amount of time that the user is blocked from providing input to the page
-- no mouse clicks, button presses, etc.
footnote:[Source and more info: https://web.dev/tbt/[web.dev]]

Number of long tasks::
Long tasks are periods of time where the main UI thread is working for 50ms or longer.
This could be because of a rerender, event handler, etc., and causes problems like delayed time to interactivity
and variable input latency.

> footnote:[Source and more info: https://web.dev/tbt/[developer.mozilla.org]]

Author: doh

// Is this right?
Longest long task duration::

> The longest of the long tasks.

Reviewer: ```suggestion
Duration of the longest long talk on the page. 
```

// Is this right?
Total long tasks duration::

> Sum duration of all long tasks.

Reviewer: ```suggestion
Sum of all long task duration that occur on the page. 
```

Reviewer: Or we can also say as `Total duration of all long tasks`

====

> These metrics tell an important story about how users may judge your website.

Reviewer: ```suggestion
These metrics tell an important story about how users may experience your website.
```

> The problem is that developers shouldn't have to become experts at interpreting and acting on metrics like paint timings,

Reviewer: May be we can rephrase this part a bit? We can call them as signals or opportunities to developers?

Author: I tried. Let me know what you think

TIP: Beginning in 2021, Google will start using core web vitals as part of their ranking algorithm,
and will open up the opportunity for websites to rank in the "top stories"
position without needing to leverage https://amp.dev/[AMP].
footnote:[Source and more info: https://webmasters.googleblog.com/2020/05/evaluating-page-experience.html[webmasters.googleblog.com]]

[discrete]
[[user-experience-distribution]]
=== Load/view distribution

Operating system, browser family, and geographic location can all have a massive impact on how visitors
experience your website.
This data can help you understand when and where your users are visiting from, and can help you
prioritize optimizations -- for example, prioritizing improvements for the most popular browsers visiting your site.

[role="screenshot"]
image::images/visitor-breakdown.png[User experience visitor breakdown]

[discrete]
[[user-experience-errors]]
=== Error breakdown

Variation in users' software and hardware makes it nearly impossible to test for every combination.
And, as JavaScript continues to get more and more complex,
the need for user experience monitoring and error reporting only increases.

> {user-experience} removes this blindspot by surfacing JavaScript errors that are

Reviewer: ```suggestion
Error monitoring removes this blindspot by surfacing JavaScript errors that are
```
?

// To do: Add a page that talks about how synthetics and UE work together. Passive/Active, etc.

[[user-experience]]

> = {user-experience}

Reviewer 2: Should this title focus on a user goal, similar to what we have for the logs/metrics/uptime topics? Something like Analyze user experience or Monitor user experience?

Reviewer 2: Looks like this has been addressed in this [PR](https://github.com/elastic/observability-docs/pull/198).

= {user-experience}

{user-experience} provides a way to quantify and analyze the perceived performance of your web application.
Unlike testing environments, {user-experience} data reflects actual users' real-world experiences.
Powered by the APM Real user monitoring (RUM) agent, all it takes is a few lines of code to begin
surfacing key user experience metrics and investigative tools.

> Drill down further by looking at data by URL, device type, operating system, browser, and location --

Reviewer 3: We don't support device type as a filter or break down yet. It's device make and model right now.

[role="screenshot"]
image::images/user-experience-tab.png[User experience tab]

[discrete]
[[why-user-experience]]
== Why is {user-experience} important?

> Search engines are placing an increasing importance on {user-experience} when organically ranking websites.

Reviewer 2: ```suggestion
Search engines are placing increasing importance on {user-experience} when organically ranking websites.
```

> Every website has goals -- some sites want users to buy a product, sign up for a mailing list, download an app,

Reviewer 3: I love the sentiment here but given that we don't support business outcome capture or sessions yet, I think we should re-focus this on Core Web Vitals instead. Here's what I sliced together for the company report email as a starting place: 

 In User Experience, we are so excited to feature Google’s Core Web Vitals which supports the increasing importance search engines are placing on user experience when organically ranking websites. Core Web Vitals are based on metrics that score three key areas of user experience: loading performance, visual stability and interactivity. This is really important because Core Web Vitals are set to become the main performance measurement in Google’s ranking factors. If your site performance is bad - you will be down-ranked! Also, if you’re a content-based site that wants to appear in the “Top Stories” section of Google’s search results, you have good Core Web Vitals results. In this case time (speed) really does equal money!

Author: Fantastic, thank you.

metrics that score three key areas of user experience: loading performance, visual stability, and interactivity.
These Core Web Vitals are set to become the main performance measurement in Google's ranking factors.

> If you’re a content-based site that wants to appear in the “Top Stories” section of Google’s search results,

Author: Neat. TIL. Thank you!

footnote:[Source and more info: https://neilpatel.com/blog/loading-time/[neilpatel.com]]
In another, Amazon calculated that a page load slowdown of just one second would cut conversions by
7% -- costing them $1.6B in sales each year.
footnote:[Source and more info: https://www.fastcompany.com/1825005/how-one-second-could-cost-amazon-16-billion-sales[fastcompany.com]]
In short, a good {user-experience} keeps your users happy and improves your website's odds of success.

[discrete]
[[how-user-experience-works]]
== How does {user-experience} work?

// I'm not even sure if this section is necessary
// Some of this is taken from the RUM docs

{user-experience} metrics are powered by the {apm-rum-ref}[APM Real User Monitoring (RUM) agent].

> The RUM agent is installed as a dependency to your application,

Reviewer 3: @vigneshshanmugam can we say "or with a few lines of javascript"? Make it seem more like the competitors in terms of ease of instrumentation?

[discrete]
[[how-user-experience-works]]
== How does {user-experience} work?

{user-experience} metrics are powered by the {apm-rum-ref}[APM Real User Monitoring (RUM) agent].
The RUM agent uses browser timing APIs, like https://w3c.github.io/navigation-timing/[Navigation Timing],
https://w3c.github.io/resource-timing/[Resource Timing], https://w3c.github.io/paint-timing/[Paint Timing],
and https://w3c.github.io/user-timing/[User Timing], to capture {user-experience}
metrics every time a user hits one of your pages.
This data is stored in {es}, where it can be visualized using {kib}.

The RUM agent can be installed as a dependency to your application, or with just a few lines of JavaScript.
It only takes a few minutes to <<instrument-apps,get started>>.

[discrete]
[[user-experience-tab]]

> == {user-experience} in {kib}

Reviewer 2: Maybe this could be a new nested page? Monitor page load durations?

Reviewer 2: Looks like this has been addressed in this [PR](https://github.com/elastic/observability-docs/pull/198).

[discrete]
[[user-experience-page-load]]
=== Page load duration

This high-level overview is your analysis starting point and answers questions like:
How long is my server taking to respond to requests?
How much time is spent parsing and painting that content?
How many page views has my site received?

You won't be able to fix any problems from viewing these metrics alone,
but you'll get a sense of the big picture as you dive deeper into your data.

[role="screenshot"]
image::images/page-load-duration.png[User experience page load duration metrics]

[discrete]
[[user-experience-metrics]]

> === {user-experience} metrics

Reviewer 2: Maybe this could be a new nested page? Analyze user experience metrics?

Reviewer 4: TBH, I would avoid adding nested pages since we will have to remove them when we move to the new system. I thought the page length was OK.

In other words, it's around this time that a user first gets feedback that the page is loading.

[role="screenshot"]
image::images/user-exp-metrics.png[User experience metrics]

// This is collapsed by default
[%collapsible]
.Metric reference
====
First contentful paint::
Focusses on the initial rendering and measures the time from when the page starts loading to when
any part of the page's content is displayed on the screen.
The agent uses the https://www.w3.org/TR/paint-timing/#first-contentful-paint[Paint timing API] available
in the browser to capture the timing information.

> FCP is captured as transaction mark for `page-load` transaction for all chromium-based browsers

Reviewer 3: I'd take out the part that says this is only in chromium based browsers as Opera supports it and Safari are about to release support as well (it's in their tech preview).

Author: Yup, this makes sense.

> .Metric reference

Author: Agreed. These should all be tooltips in the UI

Total blocking time::
The sum of the blocking time (duration above 50 ms) for each long task that occurs between the
First contentful paint and the time when the transaction is completed.
Total blocking time is a great companion metric for https://web.dev/tti/[Time to interactive]
(TTI) which is lab metric and not available in the field through browser APIs.
The agent captures TBT based on the number of long tasks that occurred during the page load lifecycle.
footnote:[More information: https://web.dev/tbt/[web.dev]]

`Long Tasks`::
A long task is any user activity or browser task that monopolize the UI thread for extended periods
(greater than 50 milliseconds) and block other critical tasks (frame rate or input latency)
from being executed.
footnote:[More information: https://developer.mozilla.org/en-US/docs/Web/API/Long_Tasks_API[developer.mozilla.org]]

Number of long tasks::
The number of long tasks.

Longest long task duration::
Duration of the longest long task on the page.

Total long tasks duration::
Total duration of all long tasks
====

These metrics tell an important story about how users may experience your website.

> The problem is that developers shouldn't have to become experts at interpreting and acting on these signals;

Author: love it

they should spend their time reacting to the opportunities that these metrics present.
For that reason (and many others), Elastic has embraced Google's Core web vitals.

[discrete]
[[user-experience-core-vitals]]
==== Core web vitals

https://web.dev/vitals/[Core web vitals] is a recent initiative from Google to introduce a new set of

> metrics that better categorize good and bad sites by quantifying the real-world {user-experience}.

Author: agreed. will fix throughout

This is done by looking at three key metrics: loading performance, visual stability, and interactivity:

[role="screenshot"]
image::images/web-dev-vitals.png[Web dev vitals (image source: https://web.dev/vitals)]

> Image source: https://web.dev/vitals/[web.dev/vitals]

Author: https://creativecommons.org/licenses/by/4.0/

Free to redistribute and use with attribution.

To provide a good user experience, Google recommends an LCP of fewer than 2.5 seconds.
footnote:[Source: https://web.dev/lcp/[web.dev]]

First input delay (FID)::
Load responsiveness. FID measures the time between a user's first interaction with a page, like a click,
and when the page can respond to those interactions.
To provide a good user experience, Google recommends a FID of less than 100 milliseconds.
footnote:[Source: https://web.dev/fid/[web.dev]]

Cumulative layout shift (CLS)::
Visual stability. Is content moving around because of async resource loading or dynamic content additions?
CLS measures these frustrating unexpected layout shifts.
To provide a good user experience, Google recommends a CLS score of less than `.1`.
footnote:[Source: https://web.dev/cls/[web.dev]]

TIP: Beginning in 2021, Google will start using core web vitals as part of their ranking algorithm
and will open up the opportunity for websites to rank in the "top stories"
position without needing to leverage https://amp.dev/[AMP].
footnote:[Source: https://webmasters.googleblog.com/2020/05/evaluating-page-experience.html[webmasters.googleblog.com]]

[discrete]
[[user-experience-distribution]]

> === Load/view distribution

Reviewer 2: Maybe this could be a new nested page?

[discrete]
[[user-experience-distribution]]
=== Load/view distribution

Operating system, browser family, and geographic location can all have a massive impact on how visitors
experience your website.
This data can help you understand when and where your users are visiting from, and can help you

> prioritize optimizations -- for example, prioritizing improvements for the most popular browsers visiting your site.

Reviewer 3: We should definitely focus on the fact that these scores will influence page rankings and also placement in top stories for content sites without requiring them to use AMP.

[role="screenshot"]
image::images/visitor-breakdown.png[User experience visitor breakdown]

[discrete]
[[user-experience-errors]]
=== Error breakdown

> Variation in users' software and hardware makes it nearly impossible to test for every combination.

Reviewer 4: I couldn't tell where you were going with the first couple of sentences. I'd try to get to the point in the first sentence, if you can, then develop it.

.Metric reference
====
First contentful paint::
Focusses on the initial rendering and measures the time from when the page starts loading to when
any part of the page's content is displayed on the screen.
The agent uses the https://www.w3.org/TR/paint-timing/#first-contentful-paint[Paint timing API] available
in the browser to capture the timing information.

> FCP is captured as transaction mark for `page-load` transaction for all chromium-based browsers

Reviewer: I would remove the RUM agent specific details from this page, Might confuse users. thoughts?

Author: Makes sense

(Chrome >60, Edge >79, Opera >47, etc.).

> footnote:[More info: https://developer.mozilla.org/en-US/docs/Glossary/First_contentful_paint[developer.mozilla.org]]

Reviewer: ++ for the footnotes and removing the RUM agent specific detail.

`Long Tasks`::
A long task is any user activity or browser task that monopolize the UI thread for extended periods
(greater than 50 milliseconds) and block other critical tasks (frame rate or input latency)
from being executed.
The agent uses the https://www.w3.org/TR/longtasks/[Long Task API], which is only available in
chromium-based browsers (Chrome >58, Edge >79, Opera >69).
Captured as `Longtask<name>` span for all managed transactions.
footnote:[More info: https://developer.mozilla.org/en-US/docs/Web/API/Long_Tasks_API[developer.mozilla.org]]

Number of long tasks::
The number of long tasks.

Longest long task duration::
Duration of the longest long task on the page.

Total long tasks duration::
Total duration of all long tasks
====

These metrics tell an important story about how users may experience your website.
The problem is that developers shouldn't have to become experts at interpreting and acting on these signals;
they should spend their time reacting to the opportunities that these metrics present.
For that reason (and many others), Elastic has embraced Google's Core web vitals.

[discrete]
[[user-experience-core-vitals]]
==== Core web vitals

> https://web.dev/vitals/[Core web vitals] is arecent initiative from Google to introduce a new set of

Reviewer: ```suggestion
https://web.dev/vitals/[Core web vitals] is a recent initiative from Google to introduce a new set of
```
